portality/view/oaipmh.py

```diff
@@ -444,42 +444,66 @@ def _parameterised_list_identifiers(dao, base_url, specified_oai_endpoint, metad
             if len(results) == 0:
                 return NoRecordsMatch(base_url)
 
-            # get the full total
+            # Get the full total
+            # Each search with a resumption token is a new search,
+            #   so the total is not the same as the first search
+            #   but is reduced by number of records already served.
+            # This full_total is the total as in the first search
             full_total = total
             if start_after is not None:
                 full_total = total + start_number - start_after[1]
 
-            # work out if we need a resumption token.  It can have one of 3 values:
-            # - None = do not include the rt in the response
-            # - some value = include in the response
-            # - the empty string = include in the response
-            resumption_token = None
-            if total > len(results):
+            # Determine where our next starting index will be
+            new_start = start_number + len(results)
+
+            # Work out if we need a resumption token.  It can have one of 3 values:
+            # - None -> do not include the rt in the response if we have a full result set
+            # - the empty string -> include in the response if this is the last set of results from an incomplete list
+            # - some value -> include in the response if there are more values to retrieve
+            if len(results) == full_total:
+                resumption_token = None
+            elif new_start == full_total:
```

Comment:
I'll be honest, I don't understand this whole method now, with the various total calculations.  If there are tests that cover all the angles, and they work, then I'm happy.

What pulled me up here is that == seems very specific, wouldn't we want to send the empty string resumption token if the new_start was >= full_total.  Maybe that can't ever happen, I can't tell from the code :)

Reply:
Good point, I'll change it to >=. If it acts in the way I think it does, it'll make no difference.